Add CRC-32 function

ggilfb opened the pull request:
Calculates CRC-32 of a binary input.

findepi commented on presto-docs/src/main/sphinx/functions/binary.rst:
```diff
@@ -61,3 +61,7 @@ Binary Functions
 
     Computes the xxhash64 hash of ``binary``.
+
+.. function:: crc32(binary) -> varbinary
```
Why the result is varbinary rather than numeric data type?
At least the commit message hints i should expect number here:

> the result of the function should be read as a long.

haozhun replied:
Hash functions all return varbinary in Presto.

We realize that crc32 is somewhat special in this manner. And I'm suggesting that this return a bigint instead.

I agree with you that the commit message shouldn't have that note if the type is varbinary. If the type is bigint, that note is not necessary.

haozhun commented:
Please change commit message to `Add crc32 function`.

In the document, please encourage the user to use ``xxhash64` instead. It's 10X faster and produces hash of better quality.

electrum commented:
Don't say "crc32 hash". Write it as "CRC-32" in the documentation and commit message. It is a checksum function, not a hash (although it can be used as a hash, but so could any function).

ggilfb commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java:
```diff
@@ -220,4 +221,20 @@ public static Slice fromHexVarbinary(@SqlType(StandardTypes.VARBINARY) Slice sli
+        // correctly in a 32 signed integer, therefore this uses a long.
+        long crc32Value = crc32.getValue();
+        hash.setLong(0, Long.reverseBytes(crc32Value));
+        return hash;
```
The function has been updated to return a long, following the comments in the pull request.

ggilfb commented on presto-docs/src/main/sphinx/functions/binary.rst:
```diff
@@ -61,3 +61,7 @@ Binary Functions
+
+.. function:: crc32(binary) -> varbinary
+
+    Computes the crc32 hash of ``binary``.
```
Updated the documentation, both returning a bigint and also encouraging people to use xxhash64.

electrum commented on presto-docs/src/main/sphinx/functions/binary.rst:
```diff
@@ -61,3 +61,9 @@ Binary Functions
+
+.. function:: crc32(binary) -> bigint
+
+    Computes CRC-32 of ``binary``.
```
```
Computes the CRC-32 of ``binary``. For general purpose hashing, use
:func:`xxhash``, as it is much faster and produces a better quality hash.

electrum commented on presto-docs/src/main/sphinx/functions/binary.rst:
```diff
@@ -61,3 +61,9 @@ Binary Functions
 
     Computes the xxhash64 hash of ``binary``.
+
+.. function:: crc32(binary) -> bigint
```
Move this above `md5` so that it is in alphabetic order

electrum replied:
(there are enough functions here to warrant separate sections, but that should be fixed separately)

electrum commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestVarbinaryFunctions.java:
```diff
@@ -214,6 +214,15 @@ public void testHashCode()
+    {
+        assertFunction("crc32(to_utf8('CRC me!'))", BIGINT, 38028046L);
+        assertFunction("crc32(to_utf8('1234567890'))", BIGINT, 639479525L);
+        assertFunction("crc32(to_utf8(CAST(1234567890 AS VARCHAR)))", BIGINT, 639479525L);
```
Add test cases for something between (2^30 - 2^31) and (2^31 - 2^32).

ggilfb replied:
Added 2 test cases for the requested ranges.

ggilfb commented:
Updated with the requested changes:
1. Squashed the commits into 1 commit.
2. Updated the documentation.
3. Added 2 test cases for the suggested ranges.

haozhun commented:
Looks good. I'm going to merge this after Travis completes. Thank you for your contribution.

haozhun commented:
Merged #8137. Thank you.
